spring-integration-file/src/main/java/org/springframework/integration/file/HeadDirectoryScanner.java

```diff
@@ -20,20 +20,40 @@
 import java.util.Arrays;
 import java.util.List;
 
+import org.springframework.integration.file.filters.CompositeFileListFilter;
 import org.springframework.integration.file.filters.FileListFilter;
 
 /**
  * A custom scanner that only returns the first <code>maxNumberOfFiles</code>
  * elements from a directory listing. This is useful to limit the number of File
- * objects in memory and therefore mutually exclusive with AcceptOnceFileListFilter.
- * 
+ * objects in memory and therefore mutually exclusive with {@code AcceptOnceFileListFilter}.
+ * It should not be used in conjunction with an {@code AcceptOnceFileListFilter}.
+ *
  * @author Iwein Fuld
+ * @author Gary Russell
  * @since 2.0
  */
 public class HeadDirectoryScanner extends DefaultDirectoryScanner {
 
+	private final HeadFilter headFilter;
+
 	public HeadDirectoryScanner(int maxNumberOfFiles) {
-		this.setFilter(new HeadFilter(maxNumberOfFiles));
+		HeadFilter headFilter = new HeadFilter(maxNumberOfFiles);
+		this.setFilter(headFilter);
+		this.headFilter = headFilter;
+	}
+
+	@Override
+	public void setFilter(FileListFilter<File> filter) {
+		if (filter instanceof CompositeFileListFilter) {
+			((CompositeFileListFilter<File>) filter).addFilter(this.headFilter);
+			super.setFilter(filter);
+		}
+		else {
+			CompositeFileListFilter<File> compositeFilter = new CompositeFileListFilter<File>();
+			compositeFilter.addFilter(filter).addFilter(this.headFilter);
```

Comment:
Seems for me the premise on the `HeadFilter` is to be the first one in the chain.
But you are breaking that rule here.

Any comments ?
Thanks